src/vos/tests/vts_dtx.c

```diff
@@ -403,10 +403,20 @@ vts_dtx_commit_visibility(struct io_test_args *args, bool ext, bool punch_obj)
 	iod.iod_size = DAOS_REC_ANY;
 
 	rc = io_test_obj_fetch(args, ++epoch, &dkey, &iod, &sgl, true);
-	assert_int_equal(rc, 0);
+	if (punch_obj) {
+		/* Object uses old punch model.  Data is visible before commit
+		 */
+		assert_int_equal(rc, 0);
 
-	/* Data record with punch DTX is readable before commit. */
-	assert_memory_equal(update_buf, fetch_buf, UPDATE_BUF_SIZE);
+		assert_memory_equal(update_buf, fetch_buf, UPDATE_BUF_SIZE);
+	} else {
+		/* Read at later timestamp than the punch should return
+		 * -DER_INPROGRESS
+		 */
+		assert_int_equal(rc, -DER_INPROGRESS);
```

Comment:
I can understand this under the new punch model. But the shortcoming is that a bad (non-completed) modification will block subsequent read that is from some non-cooperative thread. In fact, it expects the latest committed modification, not care whether there is in-processing modification or not. So return -DER_INPROGRESS seems some unreasonable.

Reply:
On some degree, the issue is that how can we abort a bad modification by force. Seems we have to use some timeout mechanism, but that is what we want to avoid in DAOS model, otherwise, it looks more like cancel ldlm lock.

Reply:
Yeah, we will need to resolve the conflict in some manner in such cases.   For now, it seems reasonable to make the reader retry as it's trying to read something that may be gone.